### PR TITLE
Initial commit replacing template first attempt to implement crypto policies

### PR DESCRIPTION
**Pull request by Jakuje**

### CODE REVIEW

File: tests/tests_reboot.yml
Comment (richm) [inline tests/tests_reboot.yml:19]: `- not crypto_policies_reboot_required`

File: tests/tests_reboot.yml
Comment (richm) [inline tests/tests_reboot.yml:8]: You don't need double quotes here - in YAML everything is a string unless you use a non-numeric string which looks like a boolean value e.g. `foo: true` - YAML will interpret this as a bool, not a string

File: tests/tests_default.yml
Comment (richm) [inline tests/tests_default.yml:22]: you don't need to use double quotes for the first and last one

File: tests/tests_reboot.yml
Comment (richm) [reply tests/tests_reboot.yml:19]: also - don't need to use double quotes for the first condition

File: vars/RedHat_8.yml
Comment (Jakuje) [inline vars/RedHat_8.yml:4]: The RHEL8 version differ only in the package list. 8.1 and 8.2 have just `crypto-policies` package, while 8.3 and newer have also `crypto-policies-scripts` (from vars/main.yml). Is if fine this way or is there a better way to arrange the files? I considered 8.0 as non-production system so I did not create this symlink, but it might make sense to create that one too for completeness, what do you think?

File: tasks/main.yml
Comment (Jakuje) [inline tasks/main.yml:73]: This is something linters complain about (comparison against `false` value) so I assume I am doing something wrong. The idea is to have a default value that will just report facts as active policy and available polcies (see readme) and if this variable is set to something else (policy name), it is applied.
Q: Is it a good idea to do something like this?
Q2: Is there a better way how to handle a varialbes that can be either undefined/false or have a set value?

File: tasks/main.yml
Comment (Jakuje) [inline tasks/main.yml:49]: This is something we would like to ideally hide inside of the `crypto-policies` scripts as it is ugly and error prone, but it is unlikely to be available in RHEL 8.4 and before so we will need an implementation here too.  Or would it make sense to write for this a python module in this role?

File: defaults/main.yml
Comment (Jakuje) [inline defaults/main.yml:6]: This is a default saying that we do not want to set policy to anything, just report what is the current and available policies (see a related comment below). The other option would be not defining it here (not recommended in best practices?) or use separate variables (more cumbersome for users?).

File: README.md
Comment (Jakuje) [inline README.md:10]: If I see right, the CI runs only CentOS. Would it be possible to add also Fedora images? Or should I set up Github Actions CI for that?

File: tests/tests_default.yml
Comment (Jakuje) [reply tests/tests_default.yml]: dropped these. Thanks

File: tests/tests_reboot.yml
Comment (Jakuje) [reply tests/tests_reboot.yml]: Dropped these. Thanks.

File: tests/tests_reboot.yml
Comment (Jakuje) [reply tests/tests_reboot.yml]: Fixed. Thanks

File: README.md
Comment (t8m) [inline README.md:3]: We use system-wide in man pages but someone from doc team should review.

File: README.md
Comment (t8m) [inline README.md:21]: `the desired crypto policy on the target system`

File: README.md
Comment (t8m) [inline README.md:22]: Maybe s/level/alone/ ? Also submodule(s)

File: README.md
Comment (t8m) [inline README.md:24]: `... if they are available on the target system ...` or maybe add it as separate sentence. `The specified base policy and policy submodule(s) must be available on the target system.`

File: README.md
Comment (t8m) [inline README.md:30]: `... updating crypto policies forces reload of some of the daemons affected by crypto policies ...`

File: README.md
Comment (t8m) [inline README.md:39]: I'd recommend defaulting this to false. Especially when you have this indicator crypto_policies_reboot_required.

Comment (Jakuje) [reply README.md:39]: I feel like we should be safe to have it on by default (it is actually rebooted only if the policy is changed).

Lets say for one-shot playbooks it will be easier for users to just set one variable and all is done without more bothering with variables.

For more complicated setups with more roles and tasks where one would like to reboot later, I think it is fine requiring the user to set one more variable `crypto_policies_reboot: false` and handle the output variable to reboot later.

File: README.md
Comment (Jakuje) [reply README.md]: Thanks. Fix in all the places.

File: README.md
Comment (Jakuje) [reply README.md]: Reworded this paragraph with your suggestions.

File: defaults/main.yml
Comment (richm) [reply defaults/main.yml:6]: I think this is fine.

File: tasks/main.yml
Comment (richm) [inline tasks/main.yml:66]: `- crypto_policies_policy | d(true)` (I'm assuming you are using `not not` to get around the "bare variables" problem - if so, use the `d(true)` filter)

File: vars/RedHat_8.yml
Comment (richm) [reply vars/RedHat_8.yml:4]: I don't think you need a symlink for 8.0
I think you should define the defaults for redhat 8.3 and later in `vars/RedHat_8.yml`, then override those defaults in `vars/RedHat_8.1.yml` and `vars/RedHat_8.2.yml`

Comment (richm) [reply vars/RedHat_8.yml:4]: ah, nevermind - I see you have defined the default `__crypto_policies_packages` in `vars/main.yml` - that's fine

File: tests/tests_update.yml
Comment (richm) [inline tests/tests_update.yml:13]: facts

File: tasks/main.yml
Comment (richm) [reply tasks/main.yml:49]: I think it is ok - but you should use the `find` module instead of using `shell` with `ls` - for this and the base policies - see https://docs.ansible.com/ansible/2.9/modules/find_module.html#find-module

File: README.md
Comment (richm) [reply README.md:10]: I'm in the process of adding crypto_policies to CI - once this is done, it will be tested on RHEL6, RHEL7, RHEL8 (8.3), RHEL X (8.4 pre-release), CentOS6, CentOS7, CentOS8, Fedora 32, and Fedora 33.

Comment (Jakuje) [reply README.md:10]: Thanks for clarification!

File: tasks/main.yml
Comment (Jakuje) [reply tasks/main.yml]: Thanks for suggestion. I did not know how to capture this more in Ansible way, but after some reading through Jinja2 filters documentation I manage to come up with nicer solution (57a9f0e3). Please, let me know if you see there some room for improvement.

File: tests/tests_update.yml
Comment (Jakuje) [reply tests/tests_update.yml]: Thanks. Fixed in current version

File: .travis.yml
Comment (richm) [reply .travis.yml:7]: > Does it make sense to run 2.10 here? if I understand right, it is the reason for the failures ...

Looks like it is missing a dependency on the python `docker` package.  For now, don't test on ansible 2.10

Comment (Jakuje) [reply .travis.yml]: Dropped that part.

**Comment by Jakuje** [comment]
@pcahyna @richm Is there anything that needs to be modified before we can merge this PR?

The (citool-staging) CI looks like having some issues -- the same tests work just fine in the (staging) CI on both combination of hosts & ansible versions.

File: README.md
Comment (richm) [inline README.md:41]: All of the other roles are "opt-in" - the role will not reboot the machine unless explicitly told to do so.  I think crypto policies should be the same - the default should be `crypto_policies_reboot: false`

File: handlers/main.yml
Comment (richm) [inline handlers/main.yml:5]: The `reboot` module takes an optional `test_command` that tells the role when the machine is at the stage where it can continue.  Is there some such command for crypto policies that you can use to determine if the system has reached the stage where the crypto policies have been applied after reboot?  Perhaps it isn't necessary in this case and you can use the default behavior of the `reboot` module.

File: README.md
Comment (Jakuje) [reply README.md:41]: In that case, we should do it the same way. Let me change it in this PR.

File: handlers/main.yml
Comment (Jakuje) [reply handlers/main.yml:5]: There is nothing special crypto policies would need to do after the reboot -- all of the changes are made before reboot and reboot just makes sure all the services and applications dynamically loading crypto libraries are started with a new configuration.

File: README.md
Comment (pcahyna) [inline README.md:27]: we use `null` as the special value to mean "do not change the current state", not `false`.

File: README.md
Comment (pcahyna) [inline README.md:37]: should be called `...reboot_ok` for consistency with a similar variable in kernel_settings: https://github.com/linux-system-roles/kernel_settings#role-variables, if the semantics is the same (it seems so).

File: README.md
Comment (pcahyna) [inline README.md:46]: what does it mean "if there are no other tasks that require playbook"?

File: defaults/main.yml
Comment (pcahyna) [reply defaults/main.yml:6]: But this default should be `null`, not `false`.

File: tasks/main.yml
Comment (pcahyna) [inline tasks/main.yml:20]: This woill silently skip hosts where the role is not supported, right? I don't think this is ok. All other roles, IIRC, fail if the distribution is not supported.

File: README.md
Comment (pcahyna) [inline README.md:12]: It is not clear what "enabling in vars" means/what the user needs to do.

File: README.md
Comment (pcahyna) [inline README.md:16]: I think that "By default" instead of "Unconfigured," would express better what you mean.

File: README.md
Comment (pcahyna) [inline README.md:56]: Have you considered calling it just `crypto_policies_available`?
On one hand, I understand that `crypto_policies_` is just a namespace prefix and the description of the variable comes after.
On the other hand, `crypto_policies_available_policies` looks silly as a needlessly repetititive repetitition.

File: README.md
Comment (Jakuje) [reply README.md:27]: Null sounds better. I did not find any note about this in the best practices nor any other examples so I wanted to use some value I know. If using `null` is preferred, should we also make a note in the best practices?

File: README.md
Comment (Jakuje) [reply README.md:37]: Probably yes. It makes it clear that it is not a force-reboot, but reboot if needed.

File: README.md
Comment (Jakuje) [reply README.md:46]: it should have said "reboot". I will reword this.

File: tasks/main.yml
Comment (Jakuje) [reply tasks/main.yml:20]: Right. I used sshd role for the inspiration here. Other, maybe better option might be using assert here so we fail if somebody accidentally runs this on unsupported os. Or is there better way to fail on unsupported hosts?

File: README.md
Comment (Jakuje) [reply README.md:12]: Does the following sound more understandable?

> The role might work elsewhere, but will require adding OS definitions in [vars/](vars/).

File: README.md
Comment (Jakuje) [reply README.md:16]: Yeah. That sounds better. I will change that.

File: README.md
Comment (Jakuje) [reply README.md:56]: The problem is that the policy name accepted as input to `crypto_policies_policy` will take not only the "bare" policies from `crypto_policies_available_policies`, but also combination of them with 1 or more `crypto_policies_available_modules`.

That is why we wanted to keep these two `available` lists similar in naming to make sure they are related to each other, rather than having one just `available` and the other about modules. Not sure if it is helpful.

**Comment by ukulekek** [comment]
[citest bad]

File: README.md
Comment (pcahyna) [inline README.md:26]: Add "The list of available base policies can be found in the `crypto_policies_available_policies` variable and the list of available policy submodules can be found in the `crypto_policies_available_modules` variable." or something similar.
Have you considered having two variables, one `crypto_policies_policy` and the other `crypto_policies_modules`, for symmetry with the output variables?

**Comment by Jakuje** [comment]
[citest bad]

File: README.md
Comment (Jakuje) [reply README.md:26]: Added.

We did not consider this as these are used together in the API of the tool. Having the use different in the crypto policies role I would consider confusing.

File: README.md
Comment (Jakuje) [reply README.md]: I changed the default to `false` after discussion with @pcahyna.

File: tasks/main.yml
Comment (pcahyna) [reply tasks/main.yml:20]: We can discuss this topic later....

File: README.md
Comment (pcahyna) [reply README.md:26]: This is certainly a valid argument.

OTOH, introducing the system role is an opportunity to create a more abstract/better structured interface, not bound to a single tool. (I say "better structured", because using a text separator like `:` to separate list items in a string is odd in YAML, as it has a list type among its primitives, so you don't need to put lists in strings and parse them.) Imagine that another tool will be used to set the crypto policy in the future, with a different UI, and this variable will just look odd. I am undecided here, @richm what do you think?

Comment (pcahyna) [reply README.md:26]: IOW, among the intended audience for a system role are administrators who neither know nor care about the UI for the underlying tool.

**Comment by ukulekek** [comment]
[citest bad]

File: README.md
Comment (richm) [reply README.md:26]: > Imagine that another tool will be used to set the crypto policy in the future, with a different UI, and this variable will just look odd. I am undecided here, @richm what do you think?

This seems to me to be a case like the `selinux` module, where it is difficult to imagine a different provider being used in the `selinux` module.  I think for the initial implementation it is ok to use a string specific to the `update-crypto-policies` command in the UI.  If sometime in the future we want to support multiple implementations, it should be easy to parse the string into whatever arguments/values are used by the new implementation, or if some new version of `update-crypto-policies` takes very different values in very different formats e.g. in RHEL9 or RHEL10.

That being said - let's keep in mind that the current `update-crypto-policies` is not the only possible implementation, and let's keep the UI as general as possible.

File: tasks/main.yml
Comment (richm) [reply tasks/main.yml:20]: Right - for now, just remove the use of `end_host` and assume there will be some other mechanism by which the user will either be able to skip this role on unsupported platforms, or otherwise handle failures gracefully.

In other system roles that have limited platform support (e.g. tlog), it is assumed that the user will structure their inventory such that hosts which do not support this functionality will be skipped.  For example, creating some sort of hostgroup e.g. `crypto_policies_hosts` so that only hosts in that group will use the role:
```yaml
- hosts: crypto_policies_hosts
  roles:
    - crypto_policies
  vars:
    crypto_policies_....
```

**Comment by jharuda** [comment]
[citest bad]

File: tasks/main.yml
Comment (Jakuje) [reply tasks/main.yml:20]: Thank you for the comments. I that case, I think it does not make sense to keep `__crypto_policies_os_supported` variable at all so I will remove it. If somebody will try to run this on unsupported host, it will probably fail on the next task, when it will try to install packages.

File: README.md
Comment (Jakuje) [reply README.md:23]: Right. The crypto policies package was put together quite recently and it is not expected to see different provider in future. I do not see any other way how the policies and subpolicies should be handled in future and therefore I would like to keep the same API in the role here, rather than reinventing something new.

The policy name (including colons) is something that is now documented all around RHEL and Fedora docs and if this role would require some artificial separation, I think it would cause more confusion than good and more issues when somebody would try to put colon separated values instead of arrays to the above arguments. I would rather keep it simple and if there will be a need for a new API, we can add alternative that will accept arrays.

**Comment by jharuda** [comment]
[citest]

File: README.md
Comment (pcahyna) [inline README.md:21]: missing space after `.`

Comment (Jakuje) [reply README.md]: Thanks. Fixed.

**Review by pcahyna — APPROVED** [review]

**Review by richm — APPROVED** [review]
lgtm

**Pull request merged**
